Basic.Azure.Storage/Extensions/BlobServiceClientExtensions.cs

```diff
@@ -0,0 +1,119 @@
+        }
+
+        public PutBlockListResponse PutBlockBlobAsList(int blockSize,
+            string containerName, string blobName, byte[] data,
+            string contentType = null, string contentEncoding = null, string contentLanguage = null, string contentMD5 = null,
+            string cacheControl = null, Dictionary<string, string> metadata = null)
+        {
+            return Task.Run(() =>
+                    PutBlockBlobAsListAsync(blockSize, containerName, blobName, data, contentType, contentEncoding, contentLanguage, contentMD5, cacheControl, metadata))
+                    .Result;
+        }
+        public async Task<PutBlockListResponse> PutBlockBlobAsListAsync(int blockSize,
+            string containerName, string blobName, byte[] data,
+            string contentType = null, string contentEncoding = null, string contentLanguage = null, string contentMD5 = null,
+            string cacheControl = null, Dictionary<string, string> metadata = null)
+        {
+            var rangesAndBlockIds = GetBlockRangesAndIds(data.Length, blockSize);
+
+            var putBlockRequests = rangesAndBlockIds
+                .Select(blockInfo => GeneratePutBlockRequestAsync(containerName, blobName, data, blockInfo))
+                .ToList();
+
+            var convertedBlockListBlockIds = rangesAndBlockIds
+                .Select(blockInfo => new BlockListBlockId { Id = blockInfo.Id, ListType = BlockListListType.Uncommitted });
+            var actualBlockIdList = new BlockListBlockIdList(convertedBlockListBlockIds);
+
+            await Task.WhenAll(putBlockRequests);
+            return await PutBlockListAsync(containerName, blobName, actualBlockIdList,
+                    cacheControl, contentType, contentEncoding, contentLanguage, contentMD5, metadata);
+        }
+
+        private static List<ArrayRangeWithBlockIdString> GetBlockRangesAndIds(int arrayLength, int blockSize)
+        {
+            var blockRangesAndIds = new List<ArrayRangeWithBlockIdString>();
+
+            var remainder = arrayLength % blockSize;
+            var startOfRemainder = arrayLength - remainder - 1;
+
+            for (var currentIndex = 0; currentIndex < arrayLength; currentIndex += blockSize)
+            {
+                blockRangesAndIds.Add(
+                    new ArrayRangeWithBlockIdString
+                    {
+                        Id = GenerateBlockId(),
+                        Offset = currentIndex,
+                        Length = (currentIndex <= startOfRemainder ? blockSize : remainder)
+                    }
+                );
+            }
+
+            return blockRangesAndIds;
+        }
+
+        private async Task<PutBlockResponse> GeneratePutBlockRequestAsync(string containerName, string blobName, byte[] fullData, ArrayRangeWithBlockIdString range)
+        {
+            var md5Task = CalculateMD5(fullData, range.Offset, range.Length);
```

Comment:
For this first version of the PutBlockBlobAsList/Intelligently, I chose to automatically calculate MD5 hashes for each block (not for the entire blob, as that is specified as part of input params).

It might be nice to eventually allow this to be disabled (very slight performance gain) and also for it to be automatically calculated for the entire blob (very slight performance hit) as overloaded methods.